File: src/detector.js
// Local color components and grouped edges. No image data leaves the browser.
export function detectRegions(rgba, w, h, originalWidth = w, originalHeight = h) {
  if (w < 16 || h < 16) return [];
  const count = w * h, colors = new Int32Array(count), edges = new Int32Array(count), result = [];
  const add = (r, method, label) => {
    const b = { x1: Math.max(0, Math.floor(r.x1 * originalWidth / w)), y1: Math.max(0, Math.floor(r.y1 * originalHeight / h)), x2: Math.min(originalWidth, Math.ceil(r.x2 * originalWidth / w)), y2: Math.min(originalHeight, Math.ceil(r.y2 * originalHeight / h)) };
    if (b.x2 - b.x1 < 12 || b.y2 - b.y1 < 10 || area(b) > originalWidth * originalHeight * .9) return;
    result.push({ bounds: b, target: { source: 'vision', label, controlType: null, automationId: null, method, originalScreenBounds: null, clipped: false } });
  };
  for (let i = 0; i < count; i++) colors[i] = ((rgba[i*4] >> 4) << 8) | ((rgba[i*4+1] >> 4) << 4) | (rgba[i*4+2] >> 4);
  components(colors, w, h, false, (r, n) => {
    const fill = n / area(r);
    if (r.x2-r.x1 >= 24 && r.y2-r.y1 >= 18 && n > 250 && (fill >= .7 || (fill >= .22 && frameSupport(colors, r, w) > .86))) add(r, 'color-region', '色块区域');
  });
  const difference = (a, b) => Math.max(Math.abs(rgba[a*4]-rgba[b*4]),Math.abs(rgba[a*4+1]-rgba[b*4+1]),Math.abs(rgba[a*4+2]-rgba[b*4+2]));
  for (let y=1; y<h-1; y++) for (let x=1; x<w-1; x++) { const i=y*w+x; if (difference(i,i+1)>36 || difference(i,i+w)>36) edges[i]=1; }
  components(dilate(edges,w,h,3,1),w,h,true,(r,n) => {
    if (r.x2-r.x1>=24 && r.y2-r.y1>=14 && n>=80) { r.x1=Math.max(0,r.x1+2); r.x2=Math.min(w,r.x2-2); add(r,'edge-region','内容区域'); }
  });
  const dedup = [];
  for (const c of result.sort((a,b)=>area(a.bounds)-area(b.bounds))) { if (!dedup.some(other=>iou(c.bounds,other.bounds)>.87)) dedup.push(c); if(dedup.length>=180)break; }
  return dedup;
}
const area = r => (r.x2-r.x1)*(r.y2-r.y1);
export function iou(a,b) { const intersection=Math.max(0,Math.min(a.x2,b.x2)-Math.max(a.x1,b.x1))*Math.max(0,Math.min(a.y2,b.y2)-Math.max(a.y1,b.y1)); return intersection/(area(a)+area(b)-intersection)||0; }
function frameSupport(colors,r,w) {
  const value=colors[r.y1*w+r.x1]; let matches=0,total=0;
  for(let x=r.x1;x<r.x2;x++){matches+=Number(colors[r.y1*w+x]===value)+Number(colors[(r.y2-1)*w+x]===value);total+=2;}
  for(let y=r.y1;y<r.y2;y++){matches+=Number(colors[y*w+r.x1]===value)+Number(colors[y*w+r.x2-1]===value);total+=2;}
  return matches/total;
}
function components(map,w,h,skipZero,found) {
  const visited=new Uint8Array(map.length),queue=new Int32Array(map.length);
  for(let start=0;start<map.length;start++){
    if(visited[start]||(skipZero&&map[start]===0))continue;
    let head=0,tail=0,value=map[start],l=start%w,r=l,t=Math.floor(start/w),b=t;
    queue[tail++]=start;visited[start]=1;
    const enqueue=i=>{if(!visited[i]&&map[i]===value){visited[i]=1;queue[tail++]=i;}};
    while(head<tail){const i=queue[head++],x=i%w,y=Math.floor(i/w);l=Math.min(l,x);r=Math.max(r,x);t=Math.min(t,y);b=Math.max(b,y);if(x>0)enqueue(i-1);if(x+1<w)enqueue(i+1);if(y>0)enqueue(i-w);if(y+1<h)enqueue(i+w);}
    found({x1:l,y1:t,x2:r+1,y2:b+1},tail);
  }
}
function dilate(input,w,h,rx,ry) {
  const integral=new Int32Array((w+1)*(h+1)),out=new Int32Array(input.length);
  for(let y=0;y<h;y++){let row=0;for(let x=0;x<w;x++){row+=input[y*w+x];integral[(y+1)*(w+1)+x+1]=integral[y*(w+1)+x+1]+row;}}
  for(let y=0;y<h;y++)for(let x=0;x<w;x++){const l=Math.max(0,x-rx),t=Math.max(0,y-ry),r=Math.min(w,x+rx+1),b=Math.min(h,y+ry+1);out[y*w+x]=Number(integral[b*(w+1)+r]-integral[t*(w+1)+r]-integral[b*(w+1)+l]+integral[t*(w+1)+l]>0);}
  return out;
}

File: app/detector.cpp
#include "detector.h"
#include <algorithm>
#include <functional>
namespace h2d {
namespace {
qint64 area(QRect r) {
    return qint64(r.width()) * r.height();
}
double overlap(QRect a, QRect b) {
    const auto intersection = area(a.intersected(b));
    return double(intersection) / std::max<qint64>(1, area(a) + area(b) - intersection);
}
void components(const QVector<int> &map, int w, int h, bool ignoreZero,
                const std::function<void(QRect, int)> &found) {
    QVector<quint8> seen(w * h);
    QVector<int> queue(w * h);
    for (int origin = 0; origin < map.size(); ++origin) {
        if (seen[origin] || (ignoreZero && !map[origin]))
            continue;
        int head = 0, tail = 1, left = origin % w, right = left, top = origin / w, bottom = top;
        queue[0] = origin;
        seen[origin] = 1;
        int value = map[origin];
        auto visit = [&](int i) {
            if (!seen[i] && map[i] == value) {
                seen[i] = 1;
                queue[tail++] = i;
            }
        };
        while (head < tail) {
            int i = queue[head++], x = i % w, y = i / w;
            left = std::min(left, x);
            right = std::max(right, x);
            top = std::min(top, y);
            bottom = std::max(bottom, y);
            if (x)
                visit(i - 1);
            if (x + 1 < w)
                visit(i + 1);
            if (y)
                visit(i - w);
            if (y + 1 < h)
                visit(i + w);
        }
        found({left, top, right - left + 1, bottom - top + 1}, tail);
    }
}
} // namespace
QVector<Candidate> detectBlocks(const QImage &original) {
    if (original.width() < 16 || original.height() < 16)
        return {};
    QImage image = original.scaled(1100, 1100, Qt::KeepAspectRatio, Qt::SmoothTransformation)
                       .convertToFormat(QImage::Format_RGB32);
    if (original.width() <= 1100 && original.height() <= 1100)
        image = original.convertToFormat(QImage::Format_RGB32);
    const int w = image.width(), h = image.height();
    QVector<int> colors(w * h), edges(w * h);
    QVector<Candidate> proposals;
    auto add = [&](QRect r, QString method, QString label) {
        int x = int(std::floor(double(r.x()) * original.width() / w)),
            y = int(std::floor(double(r.y()) * original.height() / h));
        int xx = int(std::ceil(double(r.x() + r.width()) * original.width() / w)),
            yy = int(std::ceil(double(r.y() + r.height()) * original.height() / h));
        QRect b(x, y, xx - x, yy - y);
        if (b.width() < 12 || b.height() < 10 || b == original.rect())
            return;
        auto target = manualTarget();
        target["source"] = "vision";
        target["label"] = label;
        target["method"] = method;
        proposals.append({b, target});
    };
    for (int y = 0; y < h; y++) {
        const auto row = reinterpret_cast<const QRgb *>(image.constScanLine(y));
        for (int x = 0; x < w; x++) {
            QRgb v = row[x];
            colors[y * w + x] = ((qRed(v) >> 4) << 8) | ((qGreen(v) >> 4) << 4) | (qBlue(v) >> 4);
        }
    }
    components(colors, w, h, false, [&](QRect r, int count) {
        if (r.width() < 24 || r.height() < 18 || count < 220)
            return;
        double fill = double(count) / area(r);
        int color = colors[r.y() * w + r.x()], matches = 0, border = 0;
        for (int x = r.x(); x < r.x() + r.width(); x++) {
            matches += colors[r.y() * w + x] == color;
            matches += colors[(r.y() + r.height() - 1) * w + x] == color;
            border += 2;
        }
        for (int y = r.y(); y < r.y() + r.height(); y++) {
            matches += colors[y * w + r.x()] == color;
            matches += colors[y * w + r.x() + r.width() - 1] == color;
            border += 2;
        }
        if (fill > .7 || (fill > .2 && double(matches) / border > .86))
            add(r, "color-region", "色块区域");
    });
    auto difference = [](QRgb a, QRgb b) {
        return std::max(
            {std::abs(qRed(a) - qRed(b)), std::abs(qGreen(a) - qGreen(b)), std::abs(qBlue(a) - qBlue(b))});
    };
    for (int y = 1; y < h - 1; y++) {
        auto row = reinterpret_cast<const QRgb *>(image.constScanLine(y));
        auto below = reinterpret_cast<const QRgb *>(image.constScanLine(y + 1));
        for (int x = 1; x < w - 1; x++)
            if (difference(row[x], row[x + 1]) > 36 || difference(row[x], below[x]) > 36)
                edges[y * w + x] = 1;
    }
    QVector<int> summed((w + 1) * (h + 1)), dilated(w * h);
    for (int y = 0; y < h; y++) {
        int row = 0;
        for (int x = 0; x < w; x++) {
            row += edges[y * w + x];
            summed[(y + 1) * (w + 1) + x + 1] = summed[y * (w + 1) + x + 1] + row;
        }
    }
    for (int y = 0; y < h; y++)
        for (int x = 0; x < w; x++) {
            int l = std::max(0, x - 3), r = std::min(w, x + 4), t = std::max(0, y - 1),
                b = std::min(h, y + 2);
            dilated[y * w + x] = summed[b * (w + 1) + r] - summed[t * (w + 1) + r] - summed[b * (w + 1) + l] +
                                     summed[t * (w + 1) + l] >
                                 0;
        }
    components(dilated, w, h, true, [&](QRect r, int count) {
        if (r.width() >= 24 && r.height() >= 14 && count >= 80) {
            r.adjust(2, 0, -2, 0);
            add(r, "edge-region", "内容区域");
        }
    });
    std::stable_sort(proposals.begin(), proposals.end(),
                     [](auto &a, auto &b) { return area(a.bounds) < area(b.bounds); });
    QVector<Candidate> output;
    for (const auto &p : proposals) {
        bool duplicate = false;
        for (const auto &c : output)
            if (overlap(p.bounds, c.bounds) > .87) {
                duplicate = true;
                break;
            }
        if (!duplicate)
            output.append(p);
    }
    constexpr int maxCandidates = 220;
    if (output.size() > maxCandidates) {
        QVector<Candidate> sampled;
        sampled.reserve(maxCandidates);
        // Sample the complete size range instead of dropping every large region.
        // Stable area ordering also spreads equal-sized cells across the image.
        for (int i = 0; i < maxCandidates; ++i) {
            const qsizetype index = qint64(i) * (output.size() - 1) / (maxCandidates - 1);
            sampled.append(output[index]);
        }
        output = std::move(sampled);
    }
    auto structured = detectTableBlocks(original);
    if (structured.isEmpty())
        return output;
    QVector<Candidate> other;
    for (const auto &candidate : output) {
        const bool duplicate = std::any_of(structured.cbegin(), structured.cend(), [&](const auto &table) {
            return overlap(candidate.bounds, table.bounds) > .87;
        });
        if (!duplicate)
            other.append(candidate);
    }
    const int count = std::min(int(other.size()), 480 - int(structured.size()));
    for (int i = 0; i < count; ++i)
        structured.append(other[count == 1 ? 0 : qint64(i) * (other.size() - 1) / (count - 1)]);
    return structured;
}
} // namespace h2d
